### PR TITLE
PWGHF: Adding sliceby in HF flow task

### PR DESCRIPTION
**Pull request by stefanopolitano**
This PR implements the `sliceBy` in the taskFlowCharmHadrons.cxx as discussed in https://github.com/AliceO2Group/O2Physics/pull/5089

### CODE REVIEW

**Review by fgrosa — APPROVED** [review]
Thanks @stefanopolitano! Informing also @Libeiyu

**Comment by vkucera** [comment]
Thanks @stefanopolitano

File: PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx [hunk 1]
```diff
@@ -375,17 +376,21 @@ struct HfTaskFlowCharmHadrons {
   void processDsMl(CollsWithQvecs::iterator const& collision,
                    CandDsDataWMl const& candidatesDs)
   {
-    runFlowAnalysis<DecayChannel::DsToKKPi, Partition<CandDsDataWMl>>(collision, selectedDsToKKPiWMl);
-    runFlowAnalysis<DecayChannel::DsToPiKK, Partition<CandDsDataWMl>>(collision, selectedDsToPiKKWMl);
+    auto candsDsToKKPiWMl = selectedDsToKKPiWMl->sliceByCached(aod::hf_cand::collisionId, collision.globalIndex(), cache);
+    auto candsDsToPiKKWMl = selectedDsToPiKKWMl->sliceByCached(aod::hf_cand::collisionId, collision.globalIndex(), cache);
+    runFlowAnalysis<DecayChannel::DsToKKPi, CandDsDataWMl>(collision, candsDsToKKPiWMl);
+    runFlowAnalysis<DecayChannel::DsToPiKK, CandDsDataWMl>(collision, candsDsToPiKKWMl);
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:382]:
```suggestion
    runFlowAnalysis<DecayChannel::DsToKKPi>(collision, candsDsToKKPiWMl);
    runFlowAnalysis<DecayChannel::DsToPiKK>(collision, candsDsToPiKKWMl);
```
As I commented in https://github.com/AliceO2Group/O2Physics/pull/5089, the `T1` should be deduced from the `candidates` arguments. Providing it explicitly just opens doors for bugs.

File: PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx [hunk 2]
```diff
@@ -375,17 +376,21 @@ struct HfTaskFlowCharmHadrons {
...
   // Ds with rectangular cuts
   void processDs(CollsWithQvecs::iterator const& collision,
                  CandDsData const& candidatesDs)
   {
-    runFlowAnalysis<DecayChannel::DsToKKPi, Partition<CandDsData>>(collision, selectedDsToKKPi);
-    runFlowAnalysis<DecayChannel::DsToPiKK, Partition<CandDsData>>(collision, selectedDsToPiKK);
+    auto candsDsToKKPi = selectedDsToKKPi->sliceByCached(aod::hf_cand::collisionId, collision.globalIndex(), cache);
+    auto candsDsToPiKK = selectedDsToPiKK->sliceByCached(aod::hf_cand::collisionId, collision.globalIndex(), cache);
+    runFlowAnalysis<DecayChannel::DsToKKPi, CandDsData>(collision, candsDsToKKPi);
+    runFlowAnalysis<DecayChannel::DsToPiKK, CandDsData>(collision, candsDsToPiKK);
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:393]:
```suggestion
    runFlowAnalysis<DecayChannel::DsToKKPi>(collision, candsDsToKKPi);
    runFlowAnalysis<DecayChannel::DsToPiKK>(collision, candsDsToPiKK);
```

File: PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx [hunk 3]
```diff
@@ -409,17 +414,21 @@ struct HfTaskFlowCharmHadrons {
   void processD0Ml(CollsWithQvecs::iterator const& collision,
                    CandD0DataWMl const& candidatesD0)
   {
-    runFlowAnalysis<DecayChannel::D0ToPiK, Partition<CandD0DataWMl>>(collision, selectedD0ToPiKWMl);
-    runFlowAnalysis<DecayChannel::D0ToKPi, Partition<CandD0DataWMl>>(collision, selectedD0ToKPiWMl);
+    auto candsD0ToPiKWMl = selectedD0ToPiKWMl->sliceByCached(aod::hf_cand::collisionId, collision.globalIndex(), cache);
+    auto candsD0ToKPiWMl = selectedD0ToKPiWMl->sliceByCached(aod::hf_cand::collisionId, collision.globalIndex(), cache);
+    runFlowAnalysis<DecayChannel::D0ToPiK, CandD0DataWMl>(collision, candsD0ToPiKWMl);
+    runFlowAnalysis<DecayChannel::D0ToKPi, CandD0DataWMl>(collision, candsD0ToKPiWMl);
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:420]:
```suggestion
    runFlowAnalysis<DecayChannel::D0ToPiK>(collision, candsD0ToPiKWMl);
    runFlowAnalysis<DecayChannel::D0ToKPi>(collision, candsD0ToKPiWMl);
```

File: PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx [hunk 4]
```diff
@@ -409,17 +414,21 @@ struct HfTaskFlowCharmHadrons {
...
   // D0 with rectangular cuts
   void processD0(CollsWithQvecs::iterator const& collision,
                  CandD0Data const& candidatesD0)
   {
-    runFlowAnalysis<DecayChannel::D0ToPiK, Partition<CandD0Data>>(collision, selectedD0ToPiK);
-    runFlowAnalysis<DecayChannel::D0ToKPi, Partition<CandD0Data>>(collision, selectedD0ToKPi);
+    auto candsD0ToPiK = selectedD0ToPiK->sliceByCached(aod::hf_cand::collisionId, collision.globalIndex(), cache);
+    auto candsD0ToKPi = selectedD0ToKPi->sliceByCached(aod::hf_cand::collisionId, collision.globalIndex(), cache);
+    runFlowAnalysis<DecayChannel::D0ToPiK, CandD0Data>(collision, candsD0ToPiK);
+    runFlowAnalysis<DecayChannel::D0ToKPi, CandD0Data>(collision, candsD0ToKPi);
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:431]:
```suggestion
    runFlowAnalysis<DecayChannel::D0ToPiK>(collision, candsD0ToPiK);
    runFlowAnalysis<DecayChannel::D0ToKPi>(collision, candsD0ToKPi);
```

File: PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx [hunk 5]
```diff
@@ -299,18 +300,18 @@ struct HfTaskFlowCharmHadrons {
       float massCand = 0.;
       std::vector<float> outputMl = {-999., -999.};
 
-      if constexpr (std::is_same<T1, Partition<CandDsData>>::value || std::is_same<T1, Partition<CandDsDataWMl>>::value) {
+      if constexpr (std::is_same<T1, CandDsData>::value || std::is_same<T1, CandDsDataWMl>::value) {
```
Comment (aalkin) [inline PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:303]: There is `std::is_same_v<,>` shortcut, so you do not need to add `::value`.

**Comment by stefanopolitano** [comment]
Dear @vkucera and @aalkin, thanks for your suggestions, and apologize for the late reply.
I implemented your comments, let me know if there is anything else you want me to change

**Comment by vkucera** [comment]
Please do not overwrite the PR history with force-push.

**Review by vkucera — APPROVED** [review]

**Comment by vkucera** [comment]
Otherwise looks good to me. Thanks @stefanopolitano

**Pull request merged**